include/onnxruntime/core/framework/sparse_tensor.h

```diff
@@ -1,72 +1,427 @@
+  /// This function will use Csr indices contained within the user allocated buffers.
+  /// The buffers will be used as is.
+  /// </summary>
+  /// <param name="inner_index"></param>
+  /// <param name="outer_index"></param>
+  /// <returns></returns>
+  Status UseCsrIndices(gsl::span<int64_t> inner_index, gsl::span<int64_t> outer_index);
+
+  /// <summary>
+  /// The function will allocate a single contiguous buffer and will copy values
+  /// and indices into it.
+  /// </summary>
+  /// <param name="data_transfer"></param>
+  /// <param name="data_location"></param>
+  /// <param name="values_count"></param>
+  /// <param name="values_data"></param>
+  /// <param name="inner_index"></param>
+  /// <param name="outer_index"></param>
+  /// <returns></returns>
+  Status MakeCsrData(const IDataTransfer& data_transfer,
+                     const OrtMemoryInfo& data_location,
+                     size_t values_count, void* values_data,
+                     gsl::span<const int64_t> inner_index,
+                     gsl::span<const int64_t> outer_index);
+
+  /// <summary>
+  /// Give writable access to Csr values and indices
+  /// </summary>
+  class CsrMutator {
+   public:
+    CsrMutator(Tensor& values, Tensor& inner, Tensor& outer) noexcept
+        : values_(values), inner_(inner), outer_(outer) {}
+    Tensor& Values() const noexcept { return values_; }
+    Tensor& Inner() const noexcept { return inner_; }
+    Tensor& Outer() const noexcept { return outer_; }
+
+   private:
+    std::reference_wrapper<Tensor> values_;
+    std::reference_wrapper<Tensor> inner_;
+    std::reference_wrapper<Tensor> outer_;
+  };
+
+  /// <summary>
+  /// Allocates memory for values and index and returns mutator so
+  /// data can be populated
+  /// </summary>
+  /// <param name="values_count"></param>
+  /// <param name="inner_index_count"></param>
+  /// <param name="outer_index_count"></param>
+  /// <returns></returns>
+  CsrMutator MakeCsrData(size_t values_count, size_t inner_index_count, size_t outer_index_count);
+
+  /// <summary>
+  /// Read only access to BlockSparse index
+  /// </summary>
+  class BlockSparseView {
+   public:
+    explicit BlockSparseView(const Tensor& index) noexcept
+        : index_(index) {}
+    const Tensor& Index() const noexcept { return index_; }
+
+   private:
+    std::reference_wrapper<const Tensor> index_;
+  };
+
+  /// <summary>
+  /// Return BlockSparseIndex view
+  /// </summary>
+  /// <returns></returns>
+  BlockSparseView AsBlockSparse() const;
+
+  /// <summary>
+  /// Used blocksparse index contianed in the user allocated buffer. The shape of the index
+  /// must be 2-D and must contain one tuple per each of the value blocks that
+  /// were supplied to the constructor.
+  /// </summary>
+  /// <param name="index_shape"></param>
+  /// <param name="index_data"></param>
+  /// <returns></returns>
+  Status UseBlockSparseIndices(const TensorShape& index_shape, const int32_t* index_data);
```

Comment:
>const int32_t* [](http://example.com/codeflow?start=63&length=14)

can this parameter be consistent with the other functions, e.g. UseCsrIndices() with `span<int64>`?

Reply:
This index has a complex shape, so I can't use span directly. The other indices are either 1-D or can be easily detected that they are 2-D (COO), so it would be a burden (from my actual experience)for a user to construct shape for 1-D, hence the span.

Reply:
I suppose it's a bit different in that the shape is also provided here. but why do we use int32?

Reply:
This format is to support OpenAI BlockSparse format that uses int32. Reason being is because it is used in CUDA kernels (I think).